rfd/0102-azure-node-join.md

```diff
@@ -0,0 +1,179 @@
+---
+authors: Andrew Burke (andrew.burke@goteleport.com)
+state: draft
+---
+
+# RFD 102 - Azure join method
+
+## Required Approvers
+Engineering: @jakule && @r0mant
+Product: @klizhentas && @xinding33
+Security: @reedloden
+
+## What
+
+Teleport nodes running on Azure virtual machines will be able to join a
+Teleport cluster without the need to share any secret token with the node.
+Instead, the node will present an
+[attested data document](https://learn.microsoft.com/en-us/azure/virtual-machines/linux/instance-metadata-service?tabs=linux#attested-data)
+and an access token generated from the
+[instance metadata managed identity endpoint](https://learn.microsoft.com/en-us/azure/virtual-machines/linux/instance-metadata-service?tabs=linux#managed-identity)
+to confirm the subscription the VM is running in. With a configured Azure join
+token on the auth server, the node will be allowed to join the cluster based
+on its Azure subscription.
+
+This is the Azure equivalent of EC2/IAM join as described in
+[RFD 41](https://github.com/gravitational/teleport/blob/master/rfd/0041-aws-node-join.md).
+
+## Why
+
+This will make provisioning nodes on Azure simpler and arguably more secure.
+There is no need to configure static tokens, or to manage the provisioning and
+deployment of dynamic tokens.
+
+## Details
+
+In place of a join token, nodes will present a JWT access token and an attested
+data document to the auth server, both of which will be fetched on the nodes
+via instance metadata. The attested data request can include a user-specified
+nonce which will also be signed. Like the custom headers in IAM join, the nonce
+allows Teleport to issue a challenge (crypto random string of bytes) that the
+node must include in the signed request. We can use gRPC bidirectional streams
+to implement the following flow at a new `RegisterWithAzureToken` gRPC endpoint:
+
+1. The Node initiates a `RegisterWithAzureToken` request.
+2. The auth server enforces that TLS is used for transport.
+3. The auth server sends a base64 encoded 24 byte crypto-random challenge.
+  - We use 24 bytes instead of the 32 used by the IAM method because the nonce
+    parameter is limited to 32 characters, and 24 bytes require 32 base64 characters.
+4. The node requests a) the attested data document using the challenge as the
+  nonce and b) an access token. Azure signs and returns both.
+5. The node sends a join request to the auth server, including the signed
+  document, token, and other parameters in a `RegisterUsingTokenRequest`.
+6. The auth server checks that attested data is valid:
+  - The document's pkcs7 signature is valid.
+  - The returned nonce matches the issued challenge.
+7. The auth server checks that the access token is valid:
+  - The access token's signature is valid.
+  - The access token was not issued before the start of the
+  `RegisterWithAzureToken` request.
```

Comment:
As it's not explicitly written, I'll nitpick: the auth server should also verify if the token hasn't expired